profiles/src/main/java/software/amazon/awssdk/profiles/Profile.java

```diff
@@ -0,0 +1,196 @@
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+/**
+ * A named collection of configuration stored in a {@link ProfileFile}.
+ *
+ * Special access methods are provided for loading the {@link #region()} configured in this
+ * profile. Raw property access can be made via {@link #property(String)} and {@link #properties()}.
+ *
+ * @see ProfileFile
+ */
+@SdkPublicApi
+public final class Profile implements ToCopyableBuilder<Profile.Builder, Profile> {
+
+    /**
+     * The name of this profile (minus any profile prefixes).
+     */
+    private final String name;
+
+    /**
+     * The raw properties in this profile.
+     */
+    private final Map<String, String> properties;
+
+    /**
+     * A function to resolve the profile from which this profile should derive its credentials.
+     *
+     * This is used by assume-role credentials providers to find the credentials it should use for authentication when assuming
+     * the role.
+     */
+    private final Function<String, Optional<Profile>> credentialsSourceResolver;
+
+    /**
+     * @see ProfileFile
+     * @see #builder()
+     */
+    private Profile(Builder builder) {
+        this.name = Validate.paramNotNull(builder.name, "name");
+        this.properties = Validate.paramNotNull(builder.properties, "properties");
+        this.credentialsSourceResolver = builder.credentialsSourceResolver;
+    }
+
+    /**
+     * Create a builder for defining a profile with specific attributes. For reading profiles from a file, see
+     * {@link ProfileFile}.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Retrieve the name of this profile.
+     */
+    public String name() {
+        return name;
+    }
+
+    /**
+     * Retrieve a specific raw property from this profile.
+     *
+     * @param propertyKey The name of the property to retrieve.
+     * @return The value of the property, if configured.
+     */
+    public Optional<String> property(String propertyKey) {
+        return Optional.ofNullable(properties.get(propertyKey));
+    }
+
+    /**
+     * Retrieve an unmodifiable view of all of the properties currently in this profile.
+     */
+    public Map<String, String> properties() {
+        return properties;
+    }
+
+    /**
+     * Retrieve the region for which this profile has been configured, if available.
+     */
+    public Optional<String> region() {
+        return Optional.ofNullable(properties.get(ProfileProperties.REGION));
+    }
```

Comment:
Could we just remove this now that it doesn't resolve region, and have them use property(REGION) instead? It would further decouple it from the concept of "region".

Reply:
fixed